packages/react-components/react-combobox/src/utils/useComboboxBaseState.ts

```diff
@@ -12,19 +12,27 @@ export const useComboboxBaseState = (props: ComboboxBaseProps) => {
   const { appearance = 'outline', inlinePopup = false, multiselect, onOpenChange, size = 'medium' } = props;
 
   const optionCollection = useOptionCollection();
-  const { getOptionAtIndex, getOptionById, getOptionsMatchingValue } = optionCollection;
+  const { getOptionAtIndex, getOptionsMatchingValue } = optionCollection;
 
   const [activeOption, setActiveOption] = React.useState<OptionValue | undefined>();
-  const { selectedOptions, selectOption } = useSelection(props);
 
-  // update value based on selectedOptions
+  const selectionState = useSelection(props);
+  const { selectedOptions } = selectionState;
+
+  // calculate value based on props, internal value changes, and selected options
   const isFirstMount = useFirstMount();
+  const [controllableValue, setValue] = useControllableState({
+    state: props.value,
```

Comment:
If we are going in this direction I would do something like that:

```suggestion
  const [controllableValue, setValue] = useControllableState({
    state: props.value,
    defaultState: props.defaultValue,
    initialState: multiselect ? selectedOptions.join(', ') : selectedOptions[0],
  });
```

And then call `setValue()` where it's needed and avoid the whole hook below. I remember that we have discussed it and made that hook to avoid needless renders, but now it looks even less obvious 😥

Reply:
That was actually my first thought as well 😅. The issue now is that the `useControllableState` hook can't provide the value calculation needed by Combobox on its own. Essentially it needs to be able to return to the `selectedOptions`-based value calculation when `setValue(undefined)` is called, but only if the component is uncontrolled (and even if `defaultValue` is set). 

Keeping a modified version of our previous `value` + `useMemo` calculation enables that, and should also still provide the original render optimization, even if it looks a bit wonky :D. Though if you have any thoughts on better organization, definitely let me know!

Reply:
Yeah, I initially tried to rewrite that piece and noticed that it's not so easy. If it works for you feel free to schedule something to look into it together in offline.

Reply:
will do!